File: src/main/scala/simple/treeNode/OnehundredAndTenIsBalanced.scala
package simple.treeNode

import common.TreeNodeScala

/**
  *
  * @description: 110. 平衡二叉树
  *               给定一个二叉树，判断它是否是高度平衡的二叉树。
  *               本题中，一棵高度平衡二叉树定义为：
  *               一个二叉树每个节点 的左右两个子树的高度差的绝对值不超过1。
  * @author: sunhaiting
  * @create: 2020-01-01 17:08
  **/
object OnehundredAndTenIsBalanced {
  def main(args: Array[String]): Unit = {
    val treeNode1 = new TreeNodeScala(1)
    val treeNode2 = new TreeNodeScala(2)
    val treeNode3 = new TreeNodeScala(3)
    val treeNode4 = new TreeNodeScala(4)
    val treeNode5 = new TreeNodeScala(5)
    val treeNode6 = new TreeNodeScala(6)
    val treeNode7 = new TreeNodeScala(7)
    treeNode1.right = treeNode2
    treeNode1.left = treeNode3
    treeNode3.left = treeNode4
    treeNode3.right = treeNode5
    treeNode5.left = treeNode6
    treeNode5.right = treeNode7

    val bool = isBalanced(treeNode1)
    println(bool)
  }

  /**
    * 迭代
    *
    * @param root
    * @return
    */
  def isBalanced(root: TreeNodeScala): Boolean = {
    depth(root) != -1

  }


  def depth(root: TreeNodeScala): Int = {
    if (root == null) {
      0
    } else if (depth(root.left) == -1) {
      -1
    } else if (depth(root.right) == -1) {
      -1
    } else {
      if (Math.abs(depth(root.left) - depth(root.right)) < 2) {
        Math.max(depth(root.left), depth(root.right)) + 1
      } else {
        -1
      }
    }
  }
}

File: src/main/java/simple/treeNode/OneHundredAndTenIsBalanced.java
package simple.treeNode;

import common.TreeNode;

import java.util.Stack;

/**
 * @description: 110. 平衡二叉树
 * 给定一个二叉树，判断它是否是高度平衡的二叉树。
 * 本题中，一棵高度平衡二叉树定义为：
 * 一个二叉树每个节点 的左右两个子树的高度差的绝对值不超过1。
 * @author: sunhaiting
 * @create: 2020-05-14 17:26
 **/
public class OneHundredAndTenIsBalanced {
    public static void main(String[] args) {
        TreeNode treeNode1 = new TreeNode(1);
        TreeNode treeNode2 = new TreeNode(2);
        TreeNode treeNode3 = new TreeNode(3);
        TreeNode treeNode4 = new TreeNode(4);
        TreeNode treeNode5 = new TreeNode(5);
        treeNode1.left = treeNode2;
        treeNode1.right = treeNode3;
        treeNode2.left = treeNode4;
        treeNode2.right = treeNode5;
        treeNode4.left = treeNode5;
//        treeNode3.right = treeNode4;

        boolean balanced = isBalanced(treeNode1);
        System.out.println(balanced);
    }

    /**
     * 迭代
     *
     * @param root
     * @return
     */
    public static boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {

            TreeNode node = stack.pop();
            if (node != null) {
                TreeNode left = node.left;
                TreeNode right = node.right;
                if (Math.abs(height(left) - height(right)) > 1) {
                    return false;
                }
                stack.push(left);
                stack.push(right);
            }
        }
        return true;
    }


    public static int height(TreeNode root) {
        if (root == null) return 0;
        return Math.max(height(root.left), height(root.right)) + 1;
    }


    /**
     *
     * @param root
     * @return
     */
//    public  static  boolean isBalanced2(TreeNode root){
//
//    }
}
